Source/Core/ProofObligationDescription.cs

```diff
@@ -0,0 +1,104 @@
+namespace Microsoft.Boogie;
+
+/// <summary>
+/// A multi-faceted description of a proof obligation. This class is intended
+/// to provide several forms of human-readable text summarizing the meaning of
+/// a proof obligation: one for the case where the proof has succeeded, one
+/// for the case where the proof has failed, and one for the case of quickly
+/// identifying the type of proof obligation (for inclusion in long lists of
+/// proof obligations, for example).
+/// </summary>
+public abstract class ProofObligationDescription {
+  /// <summary>
+  /// A description of what this proof obligation means when it has been
+  /// successfully proven.
+  /// </summary>
+  public abstract string SuccessDescription { get; }
+
+  /// <summary>
+  /// A description of what this proof obligation means (or might mean)
+  /// when the prover has failed to establish its validity.
+  /// </summary>
+  public virtual string FailureDescription =>
+    $"Failed to prove: {SuccessDescription}";
+
+  /// <summary>
+  /// A short description of the general type of this proof obligation.
+  /// </summary>
+  public abstract string ShortDescription { get; }
```

Comment:
Is there a commitment that this can be used as a stable identifier for this class of PO? I'm thinking about cases where languages built on top of Boogie may want to replace these more generic, Boogie-specific descriptions with more specific versions. Or is the intention that all client languages will override any such descriptions when creating the Boogie command objects?

Reply:
Using it as a stable identifier is the idea. And, yes, my intention is that languages that build on Boogie could add to or override any of these. So, when using Boogie alone, you'll have the set of stable identifiers already in this file. Dafny, for example, could keep any that it still finds useful, or could override all of them when creating command objects (as I expect Dafny would do).

Reply:
I would add a comment to that effect then. I would imagine other clients otherwise being tempted to give short sentences that are not ideal as programmatically-compared identifiers.